### ISSUE TITLE
Burrito overlay catches all events on AwesomeWM

### ISSUE THREAD

**Issue by Arkhist** [AsherGlick/Burrito#119]
I am using AwesomeWM on my machine and the Burrito overlay catches all mouse and keyboard events when placed in front of Guild Wars 2.
Is there something I'm doing wrong or is it AWM's fault?

In both cases, is there a way to fix it?

Thanks.

**Comment by AsherGlick** [comment]
DEs / WMs other then GNOME on x11 are going to all be community supported. There are some other issues that address topics similar to this but I dont think AwesomeWM specifically has been brought up. If you are using x11 then there might be a way to configure AwesomeWM to allow this, if you are using Wayland it is more difficult because wayland is still very new, despite being 14 years old, and many of the features of x11 are not yet supported in it.

**Comment by Arkhist** [comment]
Thanks for your answer, I'll look up on my end and update here if I find anything.

**Comment by ShadowJuubei** [comment]
Hello, I am having similar issues:
The overlay takes focus when clicked in its window and catches all keyboard events, mouse events are passed through to GW2.
I use Wayland, GW2 is run via lutris in xWayland, the same as Burrito.
I saw that there is Wayland support with Godot 4.3, but not sure if this fixes this issue.
(I tried opening the project in Godot 4.3, it wanted to migrate and a lot of stuff had errors and I was not sure where to start :))) )

**Comment by AsherGlick** [comment]
Don't hold your breath for wayland support in burrito, it is recommended to use x11 instead for now. The roadmap to wayland, and higher versions of godot, involves setting up vm based testing infrastructure first. I do not have finalized details of how vm configuration will work, but if you would like to help out then what you can do is provide "some scripts" to setup a VM that uses your desktop configuration.

**Comment by AsherGlick** [comment]
Action camera binds the mouse to the game so there is no click-through needed. We have seen some success on newer versions of xwayland treating this behavior properly, but I wonder if awesomewm does not handle click thorugh behavior correctly, or if it needs a setting change to work.

**Comment by Friendliness** [comment]
Yeah perhaps that's it, I can provide my awesomewm config if you'd like?
I saw you asked someone for scripts to set up their desktop environment in your VM.
I believe the basic awesomewm configuration should work too though

**Comment by DziubanMaciej** [comment]
Setting client's `focusable` field can disable all input events going to Burrito. E.g. if we do it on Burrito startup, it will never receive any inputs and all will go to Gw2.

```
client.connect_signal("manage", function (client)
    if client.name == "Burrito" then
        naughty.notify({ text = "Manage burrito" })

        client.skip_taskbar = true
        client.focusable = false
        client.ontop = true
        client:geometry{ x = 0, y = 0 }
    end
end)
```

As a dirty workaround, we may be able to hook it to some keybinding that would toggle the `focusable` field. Although, I tried it and it works a bit flaky. I think Awesome caches some focus info and simply toggling one field in runtime is not enough. Anyway, it would be good to understand the rootcause.

@AsherGlick, how does Burrito achieve the behaviour of either stealing all input events from gw2 (when UI is open) or not doing it (when playing)? What APIs are used? I'm assuming it isn't anything GNOME-specific? Are you aware of any other WMs having this problem?

EDIT: I peeked into the Burrito code, I guess `set_minimal_mouse_block` is the relevant code. It seems to be working correctly, I checked with xtruss. When UI is disabled, it calls `ShapeRectangles( ... ,  rectangles[0]={x=323, y=0, width=29, height=32})` and `ShapeRectangles( ... ,  rectangles[0]={x=0, y=0, width=2562, height=1442})` when enabled.

**Comment by coderedart** [comment]
@DziubanMaciej Behind the scenes, Burrito delegates to Godot, which uses X11's [Xshape extension](https://linux.die.net/man/3/xshape). You just give X11 a bunch of rectangles and say, in these areas, I want mouse to hit the window and in other areas, it will be transparent to mouse events. [Link to relevant godot source](https://github.com/godotengine/godot/blob/0fdbf050e0c7fc7e0a9d42c2a41ee3bfdffbd8f1/platform/linuxbsd/x11/display_server_x11.cpp#L2191)

**Comment by DziubanMaciej** [comment]
Yeah, I think that's what I saw in the xtruss logs. It looks correct. I may have to look inside awesomewm, it looks like a bug or missing feature there.

**Comment by DziubanMaciej** [comment]
I've written a simple application using `XShapeCombineRegion` and it seems to be working correctly on AwesomeWM, so the problem  is not as simple as it being not implemented in the WM. It may be a matter of an order of API calls, some window flags or some other weird X11 stuff. I'm not that experienced with X11 to know that.

Issue for AwesomeWM for tracking: https://github.com/awesomeWM/awesome/issues/4035

**Comment by DziubanMaciej** [comment]
I created a fix for AwesomeWM: https://github.com/awesomeWM/awesome/pull/4036. Currently the WM doesn't handle input shapes. It will take some time for the fix to be merged and included in an official release, so users have to build AwesomWM from source manually to get it working with burrito. The process is very straightforward, though.

There is also a problem with Burrito disappearing. I found that Guild Wars 2 sometimes requests fullscreen - `_NET_WM_STATE_FULLSCREEN` - despite "Windowed Fullscreen" setting. This happens sporadically when changing tags and clicking various windows. I've not yet figured out why the game does it, but I think it shouldn't do it. On the other hand, Burrito requests `_NET_WM_STATE_ABOVE` (because of `display/window/size/always_on_top` setting in Godot). The "fullscreen" option is considered "more on top" than "above", so Burrito disappears. @coderedart, @AsherGlick, have you observed this issue? Maybe burrito could request "fullscreen" instead of "above"?

AwesomeWM additionally has a concept of "ontop" layer that is above everything, even fullscreen windows. It can be used as a workaround. It cannot be set with EWMH, but it can be set in lua config:
```lua
client.connect_signal("manage", function (c)
    if c.name == "Godot Engine" then
        c:connect_signal("property::shape_client_input", function()
            c.ontop = true
        end)
    end
end)
```

**Comment by coderedart** [comment]
@DziubanMaciej yeah, that's a [common and hard problem](https://github.com/AsherGlick/Burrito/issues/25#issuecomment-899011256). The X11 spec *recommends* fullscreen windows stay on top of windows with above hint and most window managers (except tiling ones) follow that rule. 

The easiest workaround is to just set gw2 to window mode (with decorations like titlebar/borders removed with your window manager config) and set it to the size of your screen minus 1 pixel (eg: width = 1919 for a FHD screen). This will not trigger windowed fullscreen.

On some window managers, you can *maybe* configure gw2 window stay below all other windows. If you open gw2 on a workspace with no windows, any new windows like burrito will always remain on top of gw2.

**Comment by DziubanMaciej** [comment]
My fix to AwesomeWM has been fixed to main yesterday. Note that it isn't in any binary release, so the window manager would have to be built from source.
